client/python/openlineage/client/transport/composite.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2018-2024 contributors to the OpenLineage project
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import logging
+from functools import cached_property
+from typing import TYPE_CHECKING, Any
+
+import attr
+from openlineage.client.transport.transport import Config, Transport
+from openlineage.client.utils import get_only_specified_fields
+
+if TYPE_CHECKING:
+    from openlineage.client.client import Event
+
+log = logging.getLogger(__name__)
+
+
+@attr.s
+class CompositeConfig(Config):
+    transports: dict[str, dict[str, Any]] = attr.ib()
+    continue_on_failure: bool = attr.ib(default=True)
+
+    @classmethod
+    def from_dict(cls, params: dict[str, Any]) -> CompositeConfig:
+        if "transports" not in params:
+            msg = "composite `transports` not passed to CompositeConfig"
+            raise RuntimeError(msg)
+        return cls(**get_only_specified_fields(cls, params))
+
+
+class CompositeTransport(Transport):
+    kind = "composite"
+    config_class = CompositeConfig
+
+    def __init__(self, config: CompositeConfig) -> None:
+        self.config = config
+
+    @cached_property
+    def transports(self) -> dict[str, Transport]:
+        return self._create_transports()
+
+    def _create_transports(self) -> dict[str, Transport]:
+        from openlineage.client.transport import get_default_factory
+
+        transports = {}
+        for transport_name, transport_config in self.config.transports.items():
+            transports[transport_name] = get_default_factory().create(transport_config)
+        return transports
+
+    def emit(self, event: Event) -> None:
+        for transport_name, transport in self.transports.items():
+            try:
+                log.debug("Emitting event using %s", transport_name)
+                transport.emit(event)
+            except Exception as e:  # noqa: BLE001
+                msg = f"Transport {transport_name} failed to emit event: {e}"
```

Comment:
Please use % syntax for message strings instead of f-strings

Reply:
How would I do that both for log and exception? One requires `"msg %s" % var`, another `"msg %s", var`. f-string just seemed easier solution here.

Reply:
Instruments like Sentry use message as a key for grouping errors. Using f-string leads to generating unique messages, and grouping will not be possible.
Also logging methods are lazy, and if debug log level is disabled, then message will not be templated at all.

Reply:
Got it, I hope it works now.